### PR TITLE
New core

### PR DESCRIPTION
**Pull request by tomaka**
- Removes the `MultiaddrFuture` system from both the `Transport` and `ConnectionUpgrade` traits.
- Removes the `MuxedTransport` trait.
- Renames `Swarm` to `RawSwarm`. The name `Swarm` should be for something higher level.
- The `poll()` methods that can't fail and can't end now return directly `Async<Item>` instead of `Poll<Option<Item>, Error>`.
- The `HandledNodesTasks` and `CollectionStream` now return back the handler if they failed to reach a node. This makes it possible for the `RawSwarm` to reuse the same (yet unused) handler for multiple attempts.
- Removes the `HandlerFactory` trait. When you use the `RawSwarm` you must now manually pass a handler when you dial and on incoming connection.
- The public key mismatch situation is now handled by making the dialing future fail, instead of letting it succeed then closing the connection if necessary. This is because otherwise it would be possible to call `peer(peer_id).or_connect(handler)` and end up having `handler` handle a different peer ID, which is confusing and a possible source of bugs.
- Adds a missing `loop` in the TCP transport.

### CODE REVIEW

**Comment by tomaka** [comment]
One big issue is that this breaks floodsub, as it is no longer capable of knowing which address we're talking to.
I'm considering removing floodsub entirely and restoring it in a different PR.

File: protocols/identify/src/id_transport.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    #[inline]
+    fn nat_traversal(&self, a: &Multiaddr, b: &Multiaddr) -> Option<Multiaddr> {
+        self.transport.nat_traversal(a, b)
+    }
+}
+
+// TODO: remove unneeded bounds
+struct IdRetreiver<TMuxer>
```
Comment (twittner) [inline protocols/identify/src/id_transport.rs]: Was `id_transport.rs` supposed to be part of this PR? Regardless you probably meant `IdRetriever` here.

File: core/src/nodes/handled_node.rs [hunk 2]
```diff
@@ -22,28 +22,26 @@ use muxing::StreamMuxer;
...
 
 /// Handler for the substreams of a node.
-///
-/// > Note: When implementing the various methods, don't forget that you have to register the
-/// > task that was the latest to poll and notify it.
 // TODO: right now it is possible for a node handler to be built, then shut down right after if we
 //       realize we dialed the wrong peer for example ; this could be surprising and should either
 //       be documented or changed (favouring the "documented" right now)
-pub trait NodeHandler<TSubstream> {
+pub trait NodeHandler {
```
Comment (twittner) [inline core/src/nodes/handled_node.rs]: Why not keeping `NodeHandler` generic over the substream type?

File: protocols/identify/src/id_transport.rs (hunk 1 above)
Comment (tomaka) [reply protocols/identify/src/id_transport.rs]: I'll submit it separately.

File: core/src/nodes/handled_node.rs (hunk 2 above)
Comment (tomaka) [reply core/src/nodes/handled_node.rs]: If `TSubstream` is generic, that means you can implement the `NodeHandler` trait multiple times on the same struct with different types for `TSubstream`.

However the implementation of `NodeHandler` may need to store fields that depend on `TSubstream` and therefore it doesn't make sense to allow multiple trait implementations.

That's the theoretical explanation, in practice I've encountered several inference issues with `TSubstream` being generic.

**Comment by tomaka** [comment]
I don't see what the problem is with trailing whitespaces. Also this PR doesn't seem to add trailing whitespaces, so it can be done separately if necessary.

**Comment by twittner** [comment]
> I don't see what the problem is with trailing whitespaces.

It gets in the way by showing up in commit diffs and by potentially disturbing editing (e.g. when jumping to the end of a line). In general it requires extra work to cope with it, diff tools which need to ignore whitespace changes, extra editor commands to jump to the last non-whitespace character of a line, etc. On the other hand, what do we gain by adding trailing whitespace? Nothing.

> Also this PR doesn't seem to add trailing whitespaces

```
> git diff master --check
core/src/nodes/raw_swarm.rs:675: trailing whitespace.
+
core/src/nodes/raw_swarm.rs:1016: trailing whitespace.
+
core/src/transport/and_then.rs:82: trailing whitespace.
+
```

**Comment by tomaka** [comment]
Ah you meant whitespaces after the last character of a line. Fixed.

**Comment by jamesray1** [comment]
> One big issue is that this breaks floodsub, as it is no longer capable of knowing which address we're talking to. I'm considering removing floodsub entirely and restoring it in a different PR.

That's very problematic when I'm extending floodsub with gossipsub in #521, and they both need to be in the same file.

**Comment by jamesray1** [comment]
Now I'll need to go back to making floodsub compatible before I can continue working on gossipsub.

**Comment by jamesray1** [comment]
Do you have any suggestions on how that may be done?

File: core/src/transport/mod.rs [hunk 3]
```diff
@@ -75,19 +71,15 @@ pub trait Transport {
     /// An item should be produced whenever a connection is received at the lowest level of the
     /// transport stack. The item is a `Future` that is signalled once some pre-processing has
     /// taken place, and that connection has been upgraded to the wanted protocols.
-    type Listener: Stream<Item = Self::ListenerUpgrade, Error = IoError>;
-
-    /// Future that produces the multiaddress of the remote.
-    type MultiaddrFuture: Future<Item = Multiaddr, Error = IoError>;
```
[the remark this answers is not in the document]
Comment (jamesray1) [reply core/src/transport/mod.rs]: What's the rationale for removing `MultiaddrFuture`, considering that doing so breaks floodsub?

[the remark this answers is not in the document]
Comment (jamesray1) [reply core/src/transport/mod.rs]: Nevermind, I found https://github.com/libp2p/rust-libp2p/issues/487.

**Comment by tomaka** [comment]
> Do you have any suggestions on how that may be done?

That's unfortunately a hard question ; I'm currently waiting for a review on #573 then will probably update floodsub in a short time frame.

**Pull request merged**